File: utils/merge-rulesets.py
#!/usr/bin/env python3.6

# Merge all the .xml rulesets into a single "default.rulesets" file -- this
# prevents inodes from wasting disk space, but more importantly, this works
# around the fact that zip does not perform well on a pile of small files.

# Currently, it merges rulesets into a JSON Object for minimal overhead,
# in both storage and parsing speed.

import argparse
import glob
import json
import os
import unicodedata
import xml.etree.ElementTree

def normalize(f):
    """
    OSX and Linux filesystems encode composite characters differently in
    filenames. We should normalize to NFC: http://unicode.org/reports/tr15/
    """
    f = unicodedata.normalize("NFC", f)
    return f

# commandline arguments parsing (nobody use it, though)
parser = argparse.ArgumentParser(description="Merge rulesets")
parser.add_argument("--source_dir", default="src/chrome/content/rules")

args = parser.parse_args()

# output filename, pointed to the merged ruleset
ofn = os.path.join(args.source_dir, "default.rulesets")

# XML Ruleset Files
files = map(normalize, glob.glob(os.path.join(args.source_dir, "*.xml")))

# Under git bash, sed -i issues errors and sets the file "read-only".
if os.path.isfile(ofn):
    os.system("chmod u+w " + ofn)

# Library (JSON Object)
library = []

# Parse XML ruleset and construct JSON library
print(" * Parsing XML ruleset and constructing JSON library...")
for filename in sorted(files):
    tree = xml.etree.ElementTree.parse(filename)
    root = tree.getroot()

    ruleset = {}

    for attr in root.attrib:
        ruleset[attr] = root.attrib[attr]

    for child in root:
        if child.tag in ["target", "rule", "securecookie", "exclusion"]:
            if child.tag not in ruleset:
                ruleset[child.tag] = []
        else:
            continue

        if child.tag == "target":
            ruleset["target"].append(child.attrib["host"])

        elif child.tag == "rule":
            ru = {}
            ru["from"] = child.attrib["from"]
            ru["to"] = child.attrib["to"]

            ruleset["rule"].append(ru)

        elif child.tag == "securecookie":
            sc = {}
            sc["host"] = child.attrib["host"]
            sc["name"] = child.attrib["name"]

            ruleset["securecookie"].append(sc)

        elif child.tag == "exclusion":
            ruleset["exclusion"].append(child.attrib["pattern"])

    library.append(ruleset);

# Write to default.rulesets
print(" * Writing JSON library to %s" % ofn)
outfile = open(ofn, "w")
outfile.write(json.dumps(library))
outfile.close()

# Everything is okay.
print(" * Everything is okay.")

File: utils/merge-rulesets.js
// Merge all the .xml rulesets into a single "default.rulesets" file -- this
// prevents inodes from wasting disk space, but more importantly, works around
// the fact that zip does not perform well on a pile of small files.

// it is a translation of merge-rulesets.py (without linting, backups, git support, unicode normalisation and --fast option), but it doesn't require Python (only WSH)


var fs = new ActiveXObject("Scripting.FileSystemObject"),
shell = new ActiveXObject("WScript.Shell"),
args = WScript.Arguments,
scriptFullPath = WScript.ScriptFullName,
sourceDir = args[0],
targetFile = args[1];

var msg="";

//console or gui?
var enginefile=WScript.FullName;
var stdout=null;
if(enginefile.toLowerCase().indexOf("cscript.exe")>-1){
	stdout=WScript.StdOut;
}
function getDir(a) {
	return a.substring(0, a.lastIndexOf("\\") || a.lastIndexOf("/"))
}

//checking input

var scriptPath = getDir(WScript.ScriptFullName.toString());
if (!sourceDir) {
	sourceDir = sourceDir || (getDir(scriptPath) + "\\src\\chrome\\content\\rules");
}
var defaultFileName="default.rulesets";
if (!targetFile) {
	targetFile = sourceDir + "\\" + defaultFileName;
}

//shell.Popup(sourceDir, 0, "HTTPS Everywhere Rules Merger",64);
//shell.Popup(targetFile, 0, "HTTPS Everywhere Rules Merger",64);


if(stdout)
	stdout.WriteLine("Merging rulesets... and removing whitespaces and comments...");
var infldr = fs.GetFolder(sourceDir);
var res,fileI;

targetFile=fs.CreateTextFile(targetFile);

function clean_up(rulefile){
	//Remove extra whitespace and comments from a rulesets
	rulefile = rulefile.replace(/<!--[.\n\r\u2028]*?-->/ig,'');
	rulefile = rulefile.replace(/\s+/ig,' ');
	//rulefile = rulefile.replace(/\s*(to=|from=)/ig," $1");
	rulefile = rulefile.replace(/>\s*</, "><");
	//rulefile = rulefile.replace(/<\/ruleset>\s*/ig, "</ruleset>\n");
	rulefile = rulefile.replace(/\s*(\/>|<ruleset)/, "$1");
	return rulefile;
}

var commitId=shell.Environment("Process")("GIT_COMMIT_ID");//lol, it won't work - there is no such variable
if(commitId)
	targetFile.Write('<rulesetlibrary gitcommitid="'+commitId+'">');
else
	targetFile.Write('<rulesetlibrary>');

var fileNameRx=/\.xml$/i;
for(fileI = new Enumerator(infldr.Files); !fileI.atEnd(); fileI.moveNext()){
	if (file = fileI.item()) {
		var name = file.Name;
		if(!fileNameRx.test(name))continue;
		if(stdout)stdout.WriteLine("Processing file "+name+"...");
		var currentFileStream=file.OpenAsTextStream(1);
		var ruleset = currentFileStream.ReadAll();
		currentFileStream.Close();
		delete currentFileStream;
		ruleset = ruleset.replace("<ruleset", '<ruleset f="'+name+'"');
		targetFile.Write(clean_up(ruleset));
	}
}
targetFile.Write("</rulesetlibrary>\n");
targetFile.Close();
delete targetFile;

var currentStageMsg="Merging has been FINISHED!";

if(stdout){
	stdout.WriteLine(currentStageMsg);
}else{
	msg+="\n"+currentStageMsg;
	shell.Popup(msg, 1, "HTTPS Everywhere Rules Merger", 64);
}


/*function normalize(f){
	//OSX and Linux filesystems encode composite characters differently in filenames.
	//We should normalize to NFC: http://unicode.org/reports/tr15/.
	f = unicodedata.normalize('NFC', unicode(f, 'utf-8')).encode('utf-8')
	return f
}*/
